### PR TITLE
[clang/cas] Introduce mechanism for caching diagnostics with full source location fidelity

### PR DESCRIPTION
**Pull request by akyrtzi**
It serializes the subset of `SourceManager` state for the diagnostic source locations (`FileID`s, macro expansions, etc.) and replays diagnostics by materializing their `SourceLocation`s; that way the diagnostic consumers have full access to the include stack, expanded macros, etc.

This allows us to only cache the diagnostic information that is essential for the compilation and ignore the various ways that diagnostics can be rendered. And since diagnostic rendering runs in the context of the "local" compilation we can also remap the canonical file paths of diagnostics to their original sources.

rdar://103448355

### CODE REVIEW

**Comment by akyrtzi** [comment]
@swift-ci Please test

File: clang/tools/driver/CachedDiagnostics.cpp [hunk 1]
```diff
@@ -0,0 +1,741 @@
...
+
+std::optional<std::string>
+CachedDiagnosticSerializer::serializeEmittedDiagnostics() {
+  if (empty())
+    return std::nullopt;
+
+  SmallString<512> Buf;
+  raw_svector_ostream OS(Buf);
+  yaml::Output YOut(OS);
+  YOut << CachedDiags;
```
Comment (cachemeifyoucan) [inline clang/tools/driver/CachedDiagnostics.cpp:540]: I wonder if it is worthwhile to do a stable sort of the diagnostics before serialization. That can avoid all the indeterministic diagnostic order problem.

File: clang/test/CAS/cached-diagnostics.c [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+// RUN: %clang @%t/t1.rsp 2> %t/diags1.txt
+// RUN: diff -u %t/regular-diags1.txt %t/diags1.txt
+
+// Check that we have both the remark and source diagnostics.
+// RUN: %clang @%t/t1.rsp -Rcompile-job-cache 2> %t/diags-hit1.txt
+// RUN: FileCheck %s -input-file %t/diags-hit1.txt
+// CHECK: remark: compile job cache hit for
+// CHECK: warning: some warning
+// CHECK: warning: unknown pragma ignored
+// CHECK: warning: using the result of an assignment as a condition without parentheses
```
Comment (cachemeifyoucan) [inline clang/test/CAS/cached-diagnostics.c:21]: Should we also test driver diagnostics? Since the merging is alway a corner case.

**Review by cachemeifyoucan — APPROVED** [review]
I think eventually we should move the caching into clang driver to avoid passing all the options like prefix map to cc1 command. 

Other than that, this is mostly fine with me but I don't have enough knowledge to know what is needed to replay the diagnostics but the tests look good.

File: clang/tools/driver/CachedDiagnostics.cpp (hunk 1 above)
Comment (akyrtzi) [reply clang/tools/driver/CachedDiagnostics.cpp:540]: This is more complex than just sorting a vector. The diagnostic "notes" are implicitly associated with the non-"note" diagnostic that preceded them, so you need to keep track of "notes" separately from the other diagnostics, then order the non-"notes" diagnostics only, then re-insert the associated "notes" after the right diagnostics in the same order.

I'd like to avoid it for now, but it's something to keep in mind in case it becomes difficult to keep diagnostic emission deterministic.

File: clang/test/CAS/cached-diagnostics.c (hunk 2 above)
Comment (akyrtzi) [reply clang/test/CAS/cached-diagnostics.c:21]: Merging driver diagnostics into serialized diagnostics is tested via https://github.com/apple/llvm-project/blob/next/clang/test/CAS/fcache-compile-job-serialized-diagnostics.c#L39

**Comment by akyrtzi** [comment]
@swift-ci Please test

File: clang/lib/Frontend/CompileJobCacheKey.cpp [hunk 3]
```diff
@@ -126,6 +157,8 @@ canonicalizeForCaching(llvm::cas::ObjectStore &CAS, DiagnosticsEngine &Diags,
       FrontendOpts.DisableCachedCompileJobReplay;
   FrontendOpts.DisableCachedCompileJobReplay = false;
   FrontendOpts.IncludeTimestamps = false;
+  Opts.PathPrefixMappings = std::move(FrontendOpts.PathPrefixMappings);
+  FrontendOpts.PathPrefixMappings.clear();
```
Comment (benlangmuir) [inline clang/lib/Frontend/CompileJobCacheKey.cpp:161]: I think technically this is okay for `std::vector`, but I wonder if we should change it to

```c++
std::swap(Opts.PathPrefixMappings, FrontendOpts.PathPrefixMappings);
```

So it can't accidentally cause issues if we ever change the type to something that doesn't guarantee what its state is after a move?

File: clang/include/clang/Driver/Options.td [hunk 4]
```diff
@@ -6543,9 +6543,10 @@ def fdepscan_include_tree : Flag<["-"], "fdepscan-include-tree">,
 //
 // FIXME: Add DepscanOption flag.
 def fdepscan_prefix_map_EQ : Joined<["-"], "fdepscan-prefix-map=">,
-    Group<f_Group>, MetaVarName<"<old>=<new>">,
+    Group<f_Group>, Flags<[CC1Option]>, MetaVarName<"<old>=<new>">,
```
Comment (benlangmuir) [inline clang/include/clang/Driver/Options.td:6546]: Nit: there's a comment above on line 6484 that says these options are "driver-only".  Could we move this one to a different section, or maybe just remove that comment if we don't care about the grouping anymore?

File: clang/tools/driver/CachedDiagnostics.h [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+  /// expectation is that the buffer will be deserialized only by the same
+  /// compiler version that produced it. The format can change without
+  /// restrictions.
+  ///
+  /// The intended use is as implementation detail of compilation caching, where
+  /// the diagnostic output is associated with a compilation cache key. A
+  /// different compiler version will create different cache keys, which ensures
+  /// that the diagnostics buffer will only be read by the same compiler that
+  /// produced it.
+  Expected<std::optional<std::string>> serializeEmittedDiagnostics();
```
Comment (benlangmuir) [inline clang/tools/driver/CachedDiagnostics.h:49]: There's nothing necessarily wrong with std::string here, but I would have expected MemoryBuffer. Is there a specific reason for this choice?

**Review by benlangmuir — COMMENTED** [review]
I'd be interested to know what the performance of this is like during a cached replay when there are a lot of warnings. Is using YAML a bottleneck or does it not matter?

File: clang/tools/driver/CachedDiagnostics.h (hunk 5 above)
Comment (akyrtzi) [reply clang/tools/driver/CachedDiagnostics.h:49]: There is a slight benefit in that the `std::string` can be `std::move`d into a protobuf object.

**Comment by akyrtzi** [comment]
> I'd be interested to know what the performance of this is like during a cached replay when there are a lot of warnings. Is using YAML a bottleneck or does it not matter?

I used a test case with a file with just
```
#import <Foundation/Foundation.h>
```
and passed `-Weverything -Wsystem-headers`; this generated 3780 warnings.

when replaying diagnostics:
* deserialization of diagnostics: ~40ms
* Rendering the diagnostics: ~240ms

This indicates to me that it would be more profitable to look at improving the performance of diagnostic rendering; I'll add a `FIXME` to that effect.

File: clang/lib/Frontend/CompileJobCacheKey.cpp (hunk 3 above)
Comment (akyrtzi) [reply clang/lib/Frontend/CompileJobCacheKey.cpp:161]: Done in https://github.com/apple/llvm-project/pull/5965

File: clang/include/clang/Driver/Options.td (hunk 4 above)
Comment (akyrtzi) [reply clang/include/clang/Driver/Options.td:6546]: Removed the "only" part in https://github.com/apple/llvm-project/pull/5965
